ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Basics.hs

```diff
@@ -798,15 +789,43 @@ data ApplyMapKind' :: MapKind' -> Type -> Type -> Type where
   ApplyQueryAllMK  ::                ApplyMapKind' QueryMK' k v
   ApplyQuerySomeMK :: !(Keys k v) -> ApplyMapKind' QueryMK' k v
 
-emptyAppliedMK :: (Ord k, Eq v) => SMapKind mk -> ApplyMapKind' mk k v
-emptyAppliedMK = \case
-    SEmptyMK    -> ApplyEmptyMK
-    SKeysMK     -> ApplyKeysMK      mempty
-    SValuesMK   -> ApplyValuesMK    mempty
-    STrackingMK -> ApplyTrackingMK  mempty mempty
-    SDiffMK     -> ApplyDiffMK      mempty
-    SSeqDiffMK  -> ApplySeqDiffMK   empty
-    SQueryMK    -> ApplyQuerySomeMK mempty
+class IsApplyMapKind mk where
+  emptyMK :: forall k v. (Ord k, Eq v) => mk k v
+  default emptyMK :: forall k v. Monoid (mk k v) => mk k v
+  emptyMK = mempty
+
+  mapMK :: forall k v v'. (Ord k, Eq v, Eq v') => (v -> v') -> mk k v -> mk k v'
+  default mapMK :: forall k v v'. (Functor (mk k)) => (v -> v') -> mk k v -> mk k v'
+  mapMK = fmap
```

Comment:
Hmmm, could `IsApplyMapKind` be replaced by `Monoid` and `Functor` constraints? It's probably an over-simplification on my side but I wanted to know why.